File: sdks/java/wrapper/src/main/java/com/thunkmetrc/wrapper/RateLimiter.java
package com.thunkmetrc.wrapper;

import java.util.concurrent.Semaphore;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.Map;
import java.util.concurrent.Callable;

public class RateLimiter {
    public static class Config {
        public boolean enabled = false;
        public double maxGetPerSecondPerFacility = 50.0;
        public double maxGetPerSecondIntegrator = 150.0;
        public int maxConcurrentGetPerFacility = 10;
        public int maxConcurrentGetIntegrator = 30;
    }

    private final Config config;
    private final Semaphore integratorSem;
    private final TokenBucket integratorRate;
    private final Map<String, Semaphore> facilitySems = new ConcurrentHashMap<>();
    private final Map<String, TokenBucket> facilityRates = new ConcurrentHashMap<>();

    public RateLimiter(Config config) {
        if (config == null) config = new Config();
        this.config = config;
        this.integratorSem = new Semaphore(config.maxConcurrentGetIntegrator);
        this.integratorRate = new TokenBucket(config.maxGetPerSecondIntegrator, config.maxGetPerSecondIntegrator);
    }

    public <T> T execute(String facility, boolean isGet, Callable<T> op) throws Exception {
        if (!config.enabled || !isGet) {
            return op.call();
        }

        integratorSem.acquire();
        try {
            Semaphore facSem = null;
            if (facility != null) {
                facSem = facilitySems.computeIfAbsent(facility, k -> new Semaphore(config.maxConcurrentGetPerFacility));
                facSem.acquire();
            }

            try {
                integratorRate.acquire(1);
                if (facility != null) {
                    TokenBucket facRate = facilityRates.computeIfAbsent(facility, k -> new TokenBucket(config.maxGetPerSecondPerFacility, config.maxGetPerSecondPerFacility));
                    facRate.acquire(1);
                }

                while (true) {
                    try {
                        return op.call();
                    } catch (Exception e) {
                        if (e.getMessage() != null && e.getMessage().contains("429")) {
                            Thread.sleep(1000);
                            continue;
                        }
                        throw e;
                    }
                }
            } finally {
                if (facSem != null) facSem.release();
            }
        } finally {
            integratorSem.release();
        }
    }

    static class TokenBucket {
        private final double rate;
        private final double capacity;
        private double tokens;
        private long lastRefill;

        public TokenBucket(double rate, double capacity) {
            this.rate = rate;
            this.capacity = capacity;
            this.tokens = capacity;
            this.lastRefill = System.nanoTime();
        }

        public synchronized void acquire(double amount) throws InterruptedException {
            while (true) {
                refill();
                if (tokens >= amount) {
                    tokens -= amount;
                    return;
                }
                double missing = amount - tokens;
                long waitNs = (long) (missing / rate * 1_000_000_000.0);
                Thread.sleep(waitNs / 1_000_000, (int) (waitNs % 1_000_000));
            }
        }

        private void refill() {
            long now = System.nanoTime();
            double elapsed = (now - lastRefill) / 1_000_000_000.0;
            tokens = Math.min(capacity, tokens + elapsed * rate);
            lastRefill = now;
        }
    }
}

File: sdks/thunkmetrc-py/wrapper/src/thunkmetrc/wrapper/ratelimiter.py
import asyncio
import time
import random
import math
import re
from typing import Optional, Dict, Any, Callable

class RateLimiterConfig:
    def __init__(self, 
                 enabled: bool = False,
                 max_get_per_second_per_facility: float = 50,
                 max_get_per_second_integrator: float = 150,
                 max_concurrent_get_per_facility: int = 10,
                 max_concurrent_get_integrator: int = 30,
                 max_retries: int = 5,
                 initial_backoff: float = 0.5,
                 max_backoff: float = 10.0,
                 backoff_exponent: float = 1.5):
        self.enabled = enabled
        self.max_get_per_second_per_facility = max_get_per_second_per_facility
        self.max_get_per_second_integrator = max_get_per_second_integrator
        self.max_concurrent_get_per_facility = max_concurrent_get_per_facility
        self.max_concurrent_get_integrator = max_concurrent_get_integrator
        self.max_retries = max_retries
        self.initial_backoff = initial_backoff
        self.max_backoff = max_backoff
        self.backoff_exponent = backoff_exponent

class TokenBucket:
    def __init__(self, rate: float, capacity: float):
        self.rate = rate
        self.capacity = capacity
        self.tokens = capacity
        self.last_refill = time.monotonic()
        self.lock = asyncio.Lock()

    async def wait(self):
        async with self.lock:
            now = time.monotonic()
            elapsed = now - self.last_refill
            self.tokens = min(self.capacity, self.tokens + elapsed * self.rate)
            self.last_refill = now

            if self.tokens >= 1:
                self.tokens -= 1
                return
            
            # Wait time
            missing = 1.0 - self.tokens
            wait_time = missing / self.rate
        
        await asyncio.sleep(wait_time)
        await self.wait() # Recurse to re-check

class MetrcRateLimiter:
    def __init__(self, config: Optional[RateLimiterConfig] = None):
        if config is None:
            config = RateLimiterConfig()
        self.config = config
        
        self.integrator_rate = TokenBucket(config.max_get_per_second_integrator, config.max_get_per_second_integrator)
        self.integrator_sem = asyncio.Semaphore(config.max_concurrent_get_integrator)
        
        self.facility_rates: Dict[str, TokenBucket] = {}
        self.facility_sems: Dict[str, asyncio.Semaphore] = {}
        self.lock = asyncio.Lock()

    async def execute(self, facility: Optional[str], is_get: bool, op: Callable[[], Any]):
        if not self.config.enabled or not is_get:
            return await op()

        async with self.integrator_sem:
            async with self.get_facility_sem(facility):
                await self.integrator_rate.wait()
                if facility:
                    await self.get_facility_rate(facility).wait()
                
                # Robust Retry Loop
                retries = 0
                while True:
                    try:
                        return await op()
                    except Exception as e:
                        msg = str(e).lower()
                        # Check for 429
                        if "429" in msg or "too many requests" in msg:
                            # Try to find Retry-After in message (hacky if no header access)
                            # Assuming e might be a specialized error or just a string
                            retry_after = 0.0
                            match = re.search(r"retry-after[:\s]+(\d+)", msg)
                            if match:
                                retry_after = float(match.group(1))
                            
                            if retries >= self.config.max_retries:
                                raise e
                            
                            wait = retry_after
                            if wait <= 0:
                                # Exponential backoff + Jitter
                                backoff = self.config.initial_backoff * (self.config.backoff_exponent ** retries)
                                backoff = min(backoff, self.config.max_backoff)
                                jitter = random.uniform(0, backoff * 0.1)
                                wait = backoff + jitter
                            
                            await asyncio.sleep(wait)
                            retries += 1
                            continue
                        
                        # 5xx Errors logic could go here
                        
                        raise e

    def get_facility_rate(self, facility: str) -> TokenBucket:
        if facility not in self.facility_rates:
             self.facility_rates[facility] = TokenBucket(self.config.max_get_per_second_per_facility, self.config.max_get_per_second_per_facility)
        return self.facility_rates[facility]

    def get_facility_sem(self, facility: Optional[str]) -> asyncio.Semaphore:
        if not facility:
             return asyncio.Semaphore(9999) 
        if facility not in self.facility_sems:
             self.facility_sems[facility] = asyncio.Semaphore(self.config.max_concurrent_get_per_facility)
        return self.facility_sems[facility]
